nibabel/cifti2/cifti2.py

```diff
@@ -1345,12 +1375,18 @@ def __init__(self,
         super(Cifti2Image, self).__init__(dataobj, header=header,
                                           extra=extra, file_map=file_map)
         self._nifti_header = Nifti2Header.from_header(nifti_header)
+
         # if NIfTI header not specified, get data type from input array
         if nifti_header is None:
             if hasattr(dataobj, 'dtype'):
                 self._nifti_header.set_data_dtype(dataobj.dtype)
         self.update_headers()
 
+        if self._nifti_header.get_data_shape() != self.header.matrix.get_data_shape():
```

Comment:
We should compare the same objects we report:

```suggestion
        if self._dataobj.shape != self.header.matrix.get_data_shape():
```